galasa-managers-parent/galasa-managers-testingtools-parent/dev.galasa.selenium.manager/src/main/java/dev/galasa/selenium/internal/SeleniumEnvironment.java

```diff
@@ -0,0 +1,179 @@
+		ISeleniumManager driver;
+		Path driverRasDir = screenshotRasDirectory.resolve("driver_"+drivers.size());
+		// Get a slot or fail
+		String slotName = allocateSlot();
+		
+		try {
+			if (browser.equals(Browser.NOTSPECIFIED)) {
+				browser = Browser.valueOf(SeleniumDefaultDriver.get());
+			}
+			
+			switch(SeleniumWebDriverType.get()) {
+		    case ("local"):
+		    	driver = new LocalDriverImpl(browser, driverRasDir);
+		    	break;
+		    default:
+		    	driver =  new RemoteDriverImpl(this, seleniumManager, browser, slotName, driverRasDir);
+		    	break;
+			}
+		} catch (ConfigurationPropertyStoreException e) {
+			throw new SeleniumManagerException("Failed to fetch Driver type", e);
+		}
+		drivers.add(driver);
+		return driver;
+
+	}
+	
+	/**
+	 * A selenium specific slot to limit the number of selenium parallel runs.
+	 * 
+	 * @return
+	 * @throws SeleniumManagerException
+	 */
+	private String allocateSlot() throws SeleniumManagerException {
+		String slotKey = "driver.current.slots";
+		String slots = "";
+		String slotName = "";
+		int currentSlots = 0;
+		try {
+			slots = dss.get(slotKey);
+			if (slots != null) {
+				currentSlots = Integer.valueOf(slots);
+			}
+			
+			if (currentSlots >= SeleniumDriverMaxSlots.get()) {
+				throw new SeleniumManagerException("Failed to provsion. No slots avilable");
+			}			
+			
+			String slotNamePrefix = "SeleniumSlot_" + this.runName + "_";
+			
+			int counter = 0;
+			while ("".equals(slotName)) {
+				String slotNameAttempt = slotNamePrefix + counter;
+				if (dss.get("driver.slot." + slotNameAttempt)==null) {
+					// Found a slot name;
+					slotName = slotNameAttempt;
+					break;
+				}
+				
+				counter++;
+			}
+			
+		} catch (DynamicStatusStoreException | ConfigurationPropertyStoreException e) {
+			throw new SeleniumManagerException("Failed ot allocate slot", e);
+		}
+		
+		try {
+			dss.performActions(
+					new DssSwap(slotKey, slots, String.valueOf(currentSlots+1)),
+					new DssAdd("driver.slot." + slotName, runName)
+					);
+			this.slots.add(slotName);
+		} catch (DynamicStatusStoreException e) {
+			// Failed to set with race conditions, try again
+			allocateSlot();
+		}
+		return slotName;
+	}
+
+	/**
+	 * Screenshot all active pages on all active drivers 
+	 * 
+	 * @throws SeleniumManagerException
+	 */
+	public void screenShotPages() throws SeleniumManagerException {
+		for (ISeleniumManager driver: drivers) {
+			for (IWebPage page : driver.getPages()) {
+				page.takeScreenShot();
+			}
+		}
+	}
+
+	/**
+	 * Discard drivers 
+	 * 
+	 * @throws SeleniumManagerException
+	 */
+	public void discard() throws SeleniumManagerException {
+		for (ISeleniumManager driver: drivers) {
+			driver.discard();
+		}
+		freeSlots();
+	}
+	
+	private void freeSlots() throws SeleniumManagerException {
+		List<IDssAction> actions = new ArrayList<>();
+		try {
+			for (String slot: slots) {
```

Comment:
delete the lot and decrement the slots at the same time.    so they don't get out of sync

Reply:
This one is the same above. It forms a list of all the actions before performing all at once. There was a mistake in the number of slots to delete tho, so have resolved that